File: 2019/src/main/scala/Day1.scala
object Day1 {
  val input = scala.io.Source.fromFile("data/data1.txt", "UTF-8").getLines.toVector

  def part1(): Unit = {
    println("Answer:" + input.map(_.toInt/3 - 2).sum)
  }

  def part2(): Unit = {
    var sum = 0
    def rec(m: Int): Unit = if (m != 0) {sum += m; rec(math.max(m/3 - 2, 0))}

    input.foreach(m => rec(math.max(m.toInt/3 - 2, 0)))
    println("Answer:" + sum)
  }

  def apply() = {
    println("SOLUTION DAY 1")
    println("Running part 1")
    part1()
    println("Running part 2")
    part2()
  }
}

File: 2019/python/src/day1.py
import os
import sys
import math

def main():
    input = []
    
    with open(f'{os.path.dirname(os.path.realpath(__file__))}/../input/day1.txt') as f:
        for line in f.readlines():
            input.append(line.replace("\n",""))

    print("Part 1: " + str(part1(input)))
    print("Part 2: " + str(part2(input)))

def part1(input):
    sum = 0
    for line in input:
        mass = int(line)
        fuel = math.floor(mass/3) - 2
        sum += fuel
    return sum

def part2(input):
    sum = 0
    stack = []
    for line in input:
        stack.append(int(line))

    while stack:
        mass = stack.pop(0)
        fuel = math.floor(mass/3) - 2
        if fuel > 0:
            sum += fuel
            stack.append(fuel)
    return sum

if __name__ == "__main__":
    sys.exit(main())